ion-schema/src/isl/version_based_isl/mod.rs

```diff
@@ -0,0 +1,396 @@
+    /// Represents a reference to a named type (including aliases and built-in types)
+    Named(String, NullabilityModifier),
+    /// Represents a type reference defined as an inlined import of a type from another schema
+    TypeImport(IslImportType, NullabilityModifier),
+    /// represents an unnamed type definition reference
+    Anonymous(IslType<V>, NullabilityModifier),
+}
+
+impl<V: IslVersionTrait> IslTypeRef<V> {
+    /// Creates a named [IslTypeRef] using the name of the type referenced inside it
+    pub fn named<A: Into<String>>(name: A) -> IslTypeRef<V> {
+        IslTypeRef::Named(name.into(), NullabilityModifier::Nothing)
+    }
+
+    /// Creates a nullable [IslTypeRef] using the [IonType] referenced inside it
+    pub fn nullable_built_in(name: IonType) -> IslTypeRef<IslV1_0> {
+        IslTypeRef::Named(format!("{name}"), NullabilityModifier::Nullable)
+    }
+
+    /// Creates an anonymous [IslTypeRef] using the [IslConstraint]s referenced inside it
+    pub fn anonymous<A: Into<Vec<IslConstraint<V>>>>(constraints: A) -> IslTypeRef<V> {
+        IslTypeRef::Anonymous(
+            IslType::new(None, constraints.into()),
+            NullabilityModifier::Nothing,
+        )
+    }
+
+    /// Creates an [IslVariablyOccurringTypeRef] using the [IslConstraint]s and [Range] referenced inside it
+    pub fn variably_occurring(
+        type_ref: IslTypeRef<V>,
+        occurs: Range,
+    ) -> IslVariablyOccurringTypeRef<V> {
+        IslVariablyOccurringTypeRef::new(type_ref, occurs)
+    }
+
+    /// Creates a nullable [IslTypeRef] using the name of the type referenced inside it for ISL 2.0
+    pub fn null_or_named_type_ref<A: Into<String>>(name: A) -> IslTypeRef<IslV2_0> {
+        IslTypeRef::Named(name.into(), NullabilityModifier::NullOr)
+    }
+}
+
+/// Represents a [variably occurring type reference] that will be used by `ordered_elements` and `fields` constraints.
+///  
+/// ```ion
+/// Grammar: <VARIABLY_OCCURRING_TYPE_ARGUMENT> ::= { <OCCURS>, <CONSTRAINT>... }
+///                                      | <TYPE_ARGUMENT>
+///
+///         <OCCURS> ::= occurs: <INT>
+///            | occurs: <RANGE_INT>
+///            | occurs: optional
+///            | occurs: required
+/// ```
+///
+/// [variably occurring type reference]: https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#variably-occurring-type-arguments
+#[derive(Debug, Clone, PartialEq)]
+pub struct IslVariablyOccurringTypeRef<V: IslVersionTrait> {
+    type_ref: IslTypeRef<V>,
+    occurs: Range,
+}
+
+impl<V: IslVersionTrait> IslVariablyOccurringTypeRef<V> {
+    pub(crate) fn new(type_ref: IslTypeRef<V>, occurs: Range) -> Self {
+        Self { type_ref, occurs }
+    }
+}
+
+/// Represents an [import] in an ISL schema.
+///
+/// [import]: https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#imports
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum IslImport {
+    Schema(String),
+    Type(IslImportType),
+    TypeAlias(IslImportType),
```

Comment:
```suggestion
    Type(IslImportType),
```
`TypeAlias` is redundant since both enum variants contain the same thing.